src/TorchSharp/NN/Losses.cs

```diff
@@ -12,16 +12,56 @@ namespace TorchSharp
 {
     using Modules;
 
-    public class Loss : nn.Module
+    public abstract class Loss<T1, T2, TResult> : nn.Module<T1, T2, TResult>
     {
-        public Loss(torch.nn.Reduction reduction = nn.Reduction.Mean) : base(nameof(Loss))
+        public Loss(torch.nn.Reduction reduction = nn.Reduction.Mean) : base(nameof(Loss<T1, T2, TResult>))
         {
             this.reduction = reduction;
         }
 
         public torch.nn.Reduction reduction { get; }
     }
-    public class WeightedLoss : Loss
+
+    public abstract class Loss<T1, T2, T3, TResult> : nn.Module<T1, T2, T3, TResult>
+    {
+        public Loss(torch.nn.Reduction reduction = nn.Reduction.Mean) : base(nameof(Loss<T1, T2, T3, TResult>))
+        {
+            this.reduction = reduction;
+        }
+
+        public torch.nn.Reduction reduction { get; }
+    }
+    public abstract class Loss<T1, T2, T3, T4, TResult> : nn.Module<T1, T2, T3, T4, TResult>
+    {
+        public Loss(torch.nn.Reduction reduction = nn.Reduction.Mean) : base(nameof(Loss<T1, T2, T3, T4, TResult>))
+        {
+            this.reduction = reduction;
+        }
+
+        public torch.nn.Reduction reduction { get; }
+    }
+
+    public abstract class WeightedLoss<T1, T2, TResult> : Loss<T1, T2, TResult>
+    {
+        public WeightedLoss(Tensor? weight = null, torch.nn.Reduction reduction = nn.Reduction.Mean) : base(reduction)
+        {
+            this.weight = weight;
+        }
+
+        public Tensor? weight { get; }
+    }
+
+    public abstract class WeightedLoss<T1, T2, T3, TResult> : Loss<T1, T2, T3, TResult>
```

Comment:
Weighted loss should imho also implement `ILoss<T, TResult>`, cause the weight is optional?

Reply:
Not sure what you mean -- implementing `ILoss<T, TResult>` would mean that there's a `TResult forward(T);` and none of the loss functions have one like that.

Reply:
Sorry, I misread that one thinking it provides an optional parameter `weight` instead of having it on the class instance itself.